File: Finance API Tests/brazil_test.py
import pandas as pd
import numpy as np
import datetime as dt
from tqdm import tqdm
from joblib import Parallel, delayed
import os

'''
This code takes the price data and finds the ISINs for the securities.
It then adds some characteristics of the bonds taken from the Bloomberg
Issuance data set.
'''

dir = 'C:/Users/trmcd/OneDrive/Duke/Papers/Summer 2019 Proposal/Data/FactSet/brazil/'
p = pd.read_csv(dir + 'bond_prices_expanded.csv', index_col = 0).reset_index(drop = True)

bond_data_fn = 'C:/Users/trmcd/Dropbox/Debt Issues and Elections/Tim_Code/Output_Data/Issues/bond_issuances_py.csv'
j = pd.read_csv(bond_data_fn, index_col = 0)
j = j[(j['Country (Full Name)'] == 'BRAZIL') & (j['Maturity (Months)'] >= 0)].copy()
j = j[j['Issue Date'] >= '2002-01-01']
# isinout = pd.DataFrame(j['ISIN'].unique(), columns = ['Identifiers'])
# isinout['Trading platform (id)'] = 183 # SELIC
# isinout.iloc[:5].to_excel(dir + 'isin_list.xlsx')

# g = j[(pd.notnull(j['ISIN'])) & (j['Curr'] == 'BRL')][['ISIN', 'Bloomberg ID', 'Issuer Name', 'Security Name', 'Issue Date', 'Maturity Date']].copy().reset_index(drop = True)
g = j[(pd.notnull(j['ISIN']))][['ISIN', 'Bloomberg ID', 'Issuer Name', 'Security Name', 'Issue Date', 'Maturity Date']].copy().reset_index(drop = True)
g['stem'] = g['ISIN'].str[:-3]
map = {'BRSTNCLF1' : 'LFT',
        'BRSTNCLTN' : 'LTN',
        'BRSTNCNTB' : 'NTN-B',
        'BRSTNCNTF' : 'NTN-F',
        'BRSTNCNTC' : 'NTN-C',
        'BRSTNCNTD' : 'NTN-D'
        }
g = g[g['stem'].isin(map.keys())].copy().reset_index(drop = True)
g['type'] = [map[x] for x in g['stem']]
g = g[g['type'] != 'NTN-D'].copy().reset_index(drop = True)
g['mat_year'] = [x.split('-')[0] for x in g['Maturity Date']]
g['mat_month'] = [x.split('-')[1] for x in g['Maturity Date']]
g['mat_day'] = [x.split('-')[2] for x in g['Maturity Date']]
g[['mat_year', 'mat_month', 'mat_day']] = g[['mat_year', 'mat_month', 'mat_day']].astype('int')

def get_px_yd(asset_code):
    # asset_code = 'LFT 210104'
    # asset_code = sorted(p['asset.code'].unique())[2]
    out = p[(p['asset.code'] == asset_code)].sort_values('ref.date')

    if out.shape[0] > 0:
        if out['matur.date'].unique()[0] >= dt.datetime.strftime(dt.datetime.today(), '%Y-%m-%d'): # hasn't matured yet
            still_outstanding = 1
            days_to_mty = (pd.to_datetime(out['matur.date'].unique()[0]) - dt.datetime.today()).days
        else:
            still_outstanding = 0
            days_to_mty = (pd.to_datetime(out['matur.date'].unique()[0]) - dt.datetime.today()).days

        (issue_px, issue_yd) = out[['price.bid', 'yield.bid']].iloc[0]
        (mat_px, mat_yd) = out[['price.bid', 'yield.bid']].iloc[-1]

    else:
        (issue_px, issue_yd, mat_px, mat_yd) = (np.NaN, np.NaN, np.NaN, np.NaN)

    return (asset_code, issue_px, issue_yd, mat_px, mat_yd, still_outstanding, days_to_mty)


## figure this out: de-dupe because they have lots of taps and reissues.
ll = []
for isin in g['ISIN'].unique():
    temp = g[g['ISIN'] == isin].sort_values('Issue Date')#.iloc[0]
    temp = temp.groupby([x for x in temp.columns if x != 'Issue Date']).min().reset_index()
    if temp.shape[0] > 0:
        ll.append(temp)
g = pd.concat(ll).reset_index(drop = True)

# merge the pricing data p with the other security info g.
# p[['asset.code', 'yyqq']].drop_duplicates().sort_values(['yyqq', 'asset.code'])
datediff = pd.to_datetime(g['Maturity Date']) - pd.to_datetime(g['Issue Date'])
g['mty_length_days'] = [x.days for x in datediff]
df = p.merge(g, how = 'left',
                left_on = ['type', 'mat_year', 'mat_month', 'mat_day'],
                right_on = ['type', 'mat_year', 'mat_month', 'mat_day'])
df = df[[x for x in g.columns if x != 'Issuer Name'] + [x for x in p.columns]]

# most of the asset.codes have ISINs.
# df[pd.notnull(df['ISIN'])]['asset.code'].unique()
# df[pd.isnull(df['ISIN'])]['asset.code'].unique()

pricesandyields = Parallel(n_jobs = int(os.cpu_count()))(delayed(get_px_yd)(ac) for ac in tqdm(df['asset.code'].unique()))
maturity_info = pd.DataFrame(pricesandyields, columns = ['asset.code', 'issue_px', 'issue_yd', 'mat_px', 'mat_yd', 'still_outstanding', 'days_to_mty'])

# df contains all the price info for the available bonds and their price at maturity.
# I can then drop all the other records for the columns not of interest so I just
# have the ids and the prices at maturity.

# there are a ton fewer bonds in p than in g because it's only Brazil Treasury
# bonds observed after 2002. It's only the ones denominated in BRL.

# but for some reason it leaves out 2011 pricing data?
# df[pd.notnull(df['ISIN'])][['ISIN', 'yyqq', 'p.mean', 'y.mean', 'p.sd', 'y.sd']].drop_duplicates()

df['ref.date'] = df.loc[:,'ref.date'].astype('datetime64[ns]')
df['matur.date'] = df.loc[:,'matur.date'].astype('datetime64[ns]')
df['Maturity Date'] = df.loc[:,'Maturity Date'].astype('datetime64[ns]')
df['Issue Date'] = df.loc[:,'Issue Date'].astype('datetime64[ns]')

df_new = df.merge(maturity_info, how = 'left', on = 'asset.code')

# what's the deal here?
df[['ISIN', 'yyqq', 'p.mean', 'y.mean', 'p.sd', 'y.sd']].drop_duplicates()
# i think the problem is that some of the secuities in p don't have
# matches in g. That means they were issued 2018 or later?
df_new[['Maturity Date', 'Issue Date', 'mty_length_days']]

df_new.to_csv(dir + 'price_data_for_join_to_pairlist.csv')
'''This output goes to get_bb_data_for_brazil_2.py'''

File: Finance API Tests/brazil_test.R
# install.packages('GetTDData')
library(GetTDData) #downloads bond pricing data from the Brazilian Treasury. 
library(ggplot2)
library(lubridate)
library(dplyr)
library(zoo)

types <- c('LFT', 'LTN', 'NTN-C', 'NTN-B', 'NTN-B Principal', 'NTN-F')
asset.codes <- types   # Identifier of assets
maturity <- NULL  # Maturity date as string (ddmmyy)
# download the data. only need to do this once. 
# my.flag <- download.TD.data(asset.codes = asset.codes)
# read the files. There's a problem with NTN-C_2012.xls, so see if you can find it online? 
data <- read.TD.files(asset.codes = asset.codes, maturity = maturity)
# head(data)

data$type <- gsub(" [^ ]*$", "", data$asset.code)
data$mat_year <- sapply(strsplit(as.character(data$matur.date), '-'), "[", 1)
data$mat_month <- sapply(strsplit(as.character(data$matur.date), '-'), "[", 2)
data$mat_day <- sapply(strsplit(as.character(data$matur.date), '-'), "[", 3)
data$ttm <- sapply(difftime(time1 = data$matur.date, time2 = data$ref.date, units = "days"), "[")
data$qtm <- data$ttm / (365.25/4)
data$yymm <- gsub("-", "",gsub("[^-]*$", "", data$ref.date))
yq <- as.yearqtr(as.character(data$yymm), "%Y%m")
data$yyqq <- gsub(" ", "-", yq)

# head(data)
# hist(data$yield.bid)
# hist(data[data$typ == types[6], 'yield.bid'])

out <- data.frame()
for (ac in unique(data$asset.code)){
  out_ac <- data.frame()
  temp <- data[data$asset.code == ac,]
  temp <- temp[order(temp$ref.date),]
  issue.date <- temp[1,'ref.date']
  temp$issue.date <- issue.date
  for (dd in unique(temp$yyqq)) {
    t <- temp[temp$yyqq == dd,]
    t$p.sd <- sd(t$price.bid)
    t$p.mean <- mean(t$price.bid, na.rm = TRUE)
    t$y.sd <- sd(t$yield.bid)
    t$y.mean <- mean(t$yield.bid, na.rm = TRUE)
    out_ac <- rbind(out_ac, t)
  }
  # below normalizes the four main vars. 
  out_ac$p.mean <- (out_ac$p.mean - mean(out_ac$p.mean, na.rm = TRUE)) / 
    sd(out_ac$p.mean, na.rm = TRUE)
  out_ac$p.sd <- (out_ac$p.sd - mean(out_ac$p.sd, na.rm = TRUE)) / 
    sd(out_ac$p.sd, na.rm = TRUE)
  out_ac$y.mean <- (out_ac$y.mean - mean(out_ac$y.mean, na.rm = TRUE)) / 
    sd(out_ac$y.mean, na.rm = TRUE)
  out_ac$y.sd <- (out_ac$y.sd - mean(out_ac$y.sd, na.rm = TRUE)) / 
    sd(out_ac$y.sd, na.rm = TRUE)
  out <- rbind(out, out_ac)
}

out
unique(data[,c('asset.code', 'yyqq')])
unique(out[,c('asset.code', 'yyqq')])

out$mat_mo <- (interval((out$issue.date), (out$matur.date)) %/% months(1))
out$mat_yr <- time_length(interval(out$issue.date, out$matur.date), 'years')
out$mat_yr_short <- round(out$mat_yr)
out <- out[out$type != 'NTN-B Principal',]

# now, save it so I can export it and join it to the other stuff. 
dim(data)
dir <- 'C:/Users/trmcd/OneDrive/Duke/Papers/Summer 2019 Proposal/Data/FactSet/brazil/'
fn <- paste0(dir, 'bond_prices_expanded.csv')
fn
# data <- read.csv(fn)
# head(data)
write.csv(out, fn)


# plot data (prices)
# first, I need the spread rather than the yield. 
# all the yields drop to zero when they're maturing. This means that the 
# thing I need to move because of the hhi is the yield at maturity, which is 
# just the short-term rate, pretty much. 
# this phenomenon is very obvious in the LTNs. All the prices converge to 1k at 
# ttm==0 and disperse after that. YTM converge to something under 0.2, which 
# must be the risk-free rate for that asset. 
# this means that the thing I think is going to respond to the hhi is the 
# volatility of the price, which should be the sd of the prices? check GT. 
# the volatility risk should be something more complicated. Is there a package
# for R that allows me to price volatility risk?
subset <- out[out$type == types[2], ]
scaleFactor <- max(subset$price.bid) / max(subset$yield.bid)
# ggplot(subset, aes(x=ref.date)) +
ggplot(subset, aes(x=ttm)) +
  geom_line(aes(y=price.bid, color = asset.code)) +
  # geom_line(aes(y=yield.bid, color = asset.code)) +
  geom_line(aes(y=yield.bid * scaleFactor, color = asset.code)) +
  scale_y_continuous(name="Price", 
                     sec.axis=sec_axis(~./scaleFactor, name="Yield")
                     ) +
  # xlim(0,500) +
  theme(
    axis.title.y.left=element_text(color="blue"),
    axis.text.y.left=element_text(color="blue"),
    axis.title.y.right=element_text(color="red"),
    axis.text.y.right=element_text(color="red")
  )


## try some basic regression. 
m0 <- lm(
          # p.sd ~
          price.bid ~
           # yield.bid ~
           # asset.code +
           type +
           ttm + 
           mat_yr +
           p.sd +
           y.sd
         # data = subset
         ,data = out)
summary(m0)

hhi_data <- read.csv(paste0(dir, 'brazil_reg_data.csv'),
                    stringsAsFactors = TRUE, 
                    row.names = 1,
                    header = TRUE)
head(hhi_data)
mh <- lm(
          # price.bid ~
          p.sd ~
            # y.sd ~ 
          # yield.bid ~
          # asset.code +
          # lag(price.bid) + 
          lag(yield.bid) + 
          # issue_px +
          issue_yd +
          # mat_px +
          mat_yd + 
          still_outstanding * ttm +
          still_outstanding * days_to_mty +
          hhi +
          type +
          mat_yr #+   
          # p.sd +
          # y.sd
        # data = subset
        ,data = hhi_data)
summary(mh)

library(lme4)

# fixed intercept for the non-changing terms. 
mh <- lmer(p.sd ~
             # y.sd + 
             mat_yr + 
             price.bid + 
             # lag(p.sd) +
             # issue_px +
             # issue_yd + 
             # mat_px + 
             # mat_yd +
             # still_outstanding * days_to_mty +
             hhi +
             still_outstanding * ttm + 
             # type + 
             (1 + mat_px + mat_yd
              | type )
          # data = subset
          ,data = hhi_data)
summary(mh)
